### PR TITLE
Problem (Fix #1016): jail period/unbonding time is not verified against maximum evidence age

### LINKED ISSUE #1016: Problem: jail period/unbonding time is not verified against maximum evidence age
also https://github.com/crypto-com/chain/issues/856

### PR DESCRIPTION
**Pull request by yihuang**
Solution:
- Remove unbonding_period parameter, use max_evidence_age of consensus
  parameter instead.

### CODE REVIEW

File: client-core/src/wallet/syncer.rs [hunk 1]
```diff
@@ -728,6 +728,7 @@ mod tests {
     }
 
     #[test]
+    #[ignore]
```
Comment (devashishdxt) [inline client-core/src/wallet/syncer.rs:731]: Why is this ignored?

File: chain-abci/tests/validator_changes.rs [hunk 2]
```diff
@@ -145,20 +146,20 @@ fn check_unbonding_with_removing_validator() {
...
             // tweaking times + parameters, something more than 0.0... gets minted
-            parameters.unbonding_period = 3600 * 24 * 10000;
             parameters.rewards_config.reward_period_seconds = 365 * 24 * 3600;
             parameters.required_council_node_stake = (Coin::max() / 4).unwrap();
             parameters.rewards_config.monetary_expansion_r0 = Milli::try_new(1, 0).unwrap();
             parameters.rewards_config.monetary_expansion_tau = 1000_0000_0000_0000_0000;
             parameters.rewards_config.monetary_expansion_decay = 0;
         },
     );
+    env.max_evidence_age = 3600 * 24 * 10000;
```
Comment (devashishdxt) [inline chain-abci/tests/validator_changes.rs:162]: `unbonding_period` was in seconds and `max_evidence_age` is in nanoseconds. I'm not sure if this is correct? Check this at other places also.

Comment (yihuang) [reply chain-abci/tests/validator_changes.rs:162]: Yes, but we only take the seconds part of evidence age duration in app. This is handled in init chain event.

File: client-core/src/wallet/syncer.rs (hunk 1 above)
Comment (yihuang) [reply client-core/src/wallet/syncer.rs:731]: Because it’s not easy to update the hashes in the support files of this test case :(

Comment (yihuang) [reply client-core/src/wallet/syncer.rs:731]: I suggest ignoring this test case until we develop a convenient tool to update the support files.

Comment (tomtau) [reply client-core/src/wallet/syncer.rs:731]: https://github.com/crypto-com/chain/issues/1168

**Comment by yihuang** [comment]
> maybe a bit simpler change (non-breaking) for now will be to keep unbonding period and add validation that unbonding period == (or >= ?) max age ( in seconds instead of nanoseconds)?
> otherwise app hash changes everywhere etc.: https://travis-ci.org/github/crypto-com/chain/jobs/683687610#L1315
> breaking genesis format changes (renaming, moving or removing fields) could be done later in one go?

- I forgot to update `./ci-scripts/check-docker-app-hash.sh` which is updated now.
- I suggest ignoring the unit test(`check_wallet_syncer_app_hash_on_multiple_tx`) until we develop a convenient tool to update the supporting files.
  We can still detect breaking changes regarding app hash by `./ci-scripts/check-docker-app-hash.sh`.
- Breaking change in `CHANGELOG` is updated too.

**Comment by yihuang** [comment]
> sgx sealing test is also affected: https://github.com/crypto-com/chain/blob/master/chain-abci/src/enclave_bridge/real/test/seal.rs#L163

Done.

> 
> maybe bump the versions to 0.6.0?

We can decide the version number at the release time, #1546 can also be done in the same release.

File: chain-abci/tests/validator_changes.rs (hunk 2 above)
Comment (yihuang) [reply chain-abci/tests/validator_changes.rs:162]: https://github.com/crypto-com/chain/blob/master/chain-abci/src/app/app_init.rs#L400

**Comment by yihuang** [comment]
Fix clippy warnings, use `ChainInfo` to reduce the number of function parameters.

**Comment by yihuang** [comment]
> there's one clippy warning: https://travis-ci.org/github/crypto-com/chain/jobs/683712926#L1591
> it can potentially be ignored, but I guess this could be a good suggestion for small refactoring to group some of the parameters into new types.
> 
> e.g. these parameters
> 
> ```
>         block_time: Timespec,
>         block_height: BlockHeight,
> ```
> 
> seem to always come together.
> so there could potentially be something like:
> 
> ```
> struct BlockInfo {
>         block_time: Timespec,
>         block_height: BlockHeight,
> }
> ```
> 
> or `type BlockInfo = (Timespec, BlockHeight)`
> 
> or group even more parameters if they are always passed together.

Currently, I re-used `ChainInfo` to do this.

**Comment by tomtau** [comment]
> 
> Currently, I re-used `ChainInfo` to do this.
> In this case, it has two un-used fields out of five fields in total, and we don't need to create a new type.

I suggest not to reuse `ChainInfo`, as it is meant for tx validation (fee field is meant to be computed based on the tx size) -- otherwise, it may be error-prone, as there may be two ChainInfos in scope ("normal" one and the "fake" one with a zero fee).
Either a dedicated type for staking table usage, or extracting out needed part from ChainInfo into a new type (and nesting it in ChainInfo)

**Comment by tomtau** [comment]
Or if this refactoring that was meant to be small turns out to be big, perhaps just revert back to that original version and place allow warning annotations on those two methods with a todo comment to refactor this later

**Comment by yihuang** [comment]
Maybe just a BeginBlockInfo dedicated for begin block event for now, generalize only when there’s real need in the future.

**Pull request merged**